### ISSUE TITLE
Obscure issue with output capturing, logging module, and conftest.py

### ISSUE THREAD

**Issue by codewarrior0** [pytest-dev/pytest-xdist#4]
I normally run my test suite using the `-n <num>` option. My `setup.cfg` was previously set to pass a fixed list of paths as arguments to `py.test` via the `addopts` setting. I changed `setup.cfg` to no longer use a fixed list of paths (instead using the `python_files` glob option) in order to allow paths to be specified normally. After this change, I found that output from the `logging` module was no longer captured.

The logging module is initialized with `basicConfig` upon import of a `log` module inside the application package. This is unavoidable. The `log` module is indirectly imported by a conftest.py module. This is not unavoidable, just very inconvenient to work around; besides, I *want* to see the log output along with the results of failing tests.

Further investigation yielded some confusing results.

--

To reproduce the error, I created a test suite structured like this:

```
tests/
  foo/
    conftest.py
    test_foo.py
 bar/
    test_bar.py
```

No `setup.cfg` or similar is present.

`foo/conftest.py` initializes the `logging` module using `basicConfig`. Both `foo/test_foo.py` and `bar/test_bar.py` contain a test function that calls the `logging` module to log a single message, "Hello World". Under certain conditions, the "Hello World" message is not captured for any test, but sent to stderr instead. I have narrowed down those conditions to these invocations of `py.test` with capturing results:

- `py.test` *without* `-n <num>`, *regardless of other arguments*: Captured
- `py.test -n 3` *without other arguments*: __Not captured__
- `py.test -n 3 tests`: __Not captured__
- `py.test -n 3 tests/foo`: Captured
- `py.test -n 3 tests/foo/test_foo.py`: Captured
- `py.test -n 3 tests/bar`: Captured
- `py.test -n 3 tests/bar/test_two.py`: Captured

**Comment by codewarrior0** [comment]
Oh, I almost forgot.

I am running Windows 7, and starting pytest using a MinGW Bash shell. The issue happens on multiple Python versions and both the released versions and git versions of pytest and pytest-xdist. It also happens on an AppVeyor instance running Windows 8 on multiple Python versions and the release versions of pytest and pytest-xdist. (pytest-2.7.2, pytest-xdist-1.13.1)

**Comment by RonnyPfannschmidt** [comment]
that one will be hard to debug, we might be unable to address this before pytest-xdist 2.0 due to limited manpower (e.E. almost entirely volunteer work)

**Comment by codewarrior0** [comment]
I understand. I'll keep posting my findings in case anybody has any insight to add.

I have a hunch that it is related to the SysCapture method. When a `logging` StreamHandler is created, it stores a reference to the output stream (`sys.stderr`) by default, so SysCapture's redirection method has no effect. I'm using the default capturing method (FDCapture) but this seems to also use SysCapture at the same time.

When running `py.test -n 3 tests`, where the output is not captured, `sys.stderr` is a `_pytest.capture.EncodedFile` with a fileno==12, and the StreamHandler attached to the root logger  is a `fdopen` with a fileno==6.

When running `py.test -n 3 tests/foo`, where the output *is* captured, both `sys.stderr` and the StreamHandler are both `_pytest.capture.EncodedFile` with fileno==12.

**Comment by codewarrior0** [comment]
Actually, that's a good question. Why *does* FDCapture call SysCapture?

**Comment by codewarrior0** [comment]
> Actually, that's a good question. Why *does* FDCapture call SysCapture?
Because sys.stderr may not be a `file` with fileno==2 and *you will want to capture it anyway*.

Aha! The `fdopen` with fileno==6 above is the original `sys.stderr` created by `execnet` (via `xdist`) in the startup code run by the slave process. So I infer that in the "good" invocation, `conftest.py` is run under output capturing, so the StreamHandler stores the fileobj for the redirect; but in the "bad" invocation, `conftest.py` is *not* run under output capturing, so the StreamHandler stores the original fileobj that was in `sys.stderr`

Because of the way `execnet` works, redirecting fileno==2 at the FD level doesn't affect the StreamHandler either way. (And on that note, redirecting fileno==1 at the FD level will break `execnet`'s ability to send output back to the master process for the duration of the output capture. Worrying.)

**Comment by codewarrior0** [comment]
The `capture` plugin explicitly runs the *initial* conftest.py under output capturing, via the `pytest_load_initial_conftests` hook. 

However, conftest.py files can be found later, most likely via the `FSHookProxy` object created in `Session.gethookproxy`. The conftest files are loaded at the time the `FSHookProxy` is created, and no capturing is active at this time.

**Comment by codewarrior0** [comment]
> The logging module is initialized with basicConfig upon import of a log module inside the application package. This is unavoidable.

Whoops! This wasn't unavoidable; in fact, it was just plain dumb. I moved the basicConfig call into the function that checks for the `optparse` option for setting the logging level and the problem went away. So that's my workaround for now.

**Comment by RonnyPfannschmidt** [comment]
Fabulous work on the investigation side!!

Am I correctly assuming that it works for you now?

**Comment by codewarrior0** [comment]
Yeah, I guess. I got to the bottom of it and came up with the "right" workaround to use in my code. This is still technically an issue, but it has nothing to do with `xdist` as it turned out - it was only exposed because `execnet` has a special case for replacing `sys.stderr` on Windows.

The core issue is that `conftest.py` *may or may not* be executed with output capturing enabled, depending on whether the args passed to `pytest` indicate that the particular `conftest.py` is the *initial* conftest or not. This, combined with the `logging` module storing a reference to `sys.stderr`, *and* `execnet` replacing `sys.stderr` with a file whose fileno is not 2, causes logging output to not be captured.
